akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -77,6 +77,27 @@ private[http] final class MasterServerTerminator(log: LoggingAdapter) extends Se
     }
   }
 
+  /**
+   * Removes a previously registered per-connection terminator.
+   * Terminators must remove themselves like this once their respective connection is closed,
+   * otherwise they would leak and remain in the set indefinitely.
+   *
+   * If called with terminator that is not registered, false is returned but it is not an error.
+   *
+   * @return true if the terminator has been successfully removed.
+   */
+  @tailrec def removeConnection(terminator: ServerTerminator)(implicit ec: ExecutionContext): Boolean = {
+    terminators.get() match {
+      case v @ AliveConnectionTerminators(ts) ⇒
+        terminators.compareAndSet(v, v.copy(ts = ts - terminator)) ||
+          removeConnection(terminator) // retry
+
+      case Terminating(deadline) ⇒
+        terminator.terminate(deadline.timeLeft)(ec)
```

Comment:
I think this line needs to be removed - we don't want to trigger this terminator when removing it, right?

Reply:
Hm, I had thought it's fine, but you're right I think, here's the problem case which confirms this should be removed I guess:

- remove is issued when the connection stream completes/fails.
- the phase becomes `Terminating()` when terminate is issued on the master terminator
- this is when the user calls terminate on the server binding
- existing connections are still alive, the terminate() call results in all existing connections be called terminate on; this is a CAS into the `Terminating(deadline)` container; val connectionsTerminated = Future.sequence(ts.map { t ⇒ t.terminate(timeout) })
- if they remain alive and the deadline triggers, the the stream is killed and the terminate enters here, with an terminator that already was called on with terminate... this is a 2nd call on the same instance and should not happen. -- you're right, removing this invocation here